### DISCUSSION TITLE
Effects for Lightbulb

### DISCUSSION THREAD

**Discussion by DukemFIVE** [HomeSpan/HomeSpan#458] [General]
Hey is there a way to include this menu in HomeSpan ?  Is a Text selection for lights.
![image](https://user-images.githubusercontent.com/119198285/206919246-7234244d-8def-4dae-a238-2ce10d37a037.png)

Its avalible in Homebridge simple wled plugin. I think they added it with:
effectInputSource = this.wledAccessory.addService(this.hap.Service.InputSource, effectID, effectName);

Would be really cool for led effects..

**Comment by ghost** [comment]
Absolutely love the idea.
I am really new to Homespan, but it looks like in the Homebridge plugin it is using a Lightbulb using a TV-Service.
Guess you could try to use "Television.ino" from the "Other Examples" as a base.
So I would think this is possible.
Currently I am a bit busy, but should be able to try it out on the weekend

**Comment by HomeSpan** [comment]
I tried different combination of including InputSource along with a Lightbulb was not able to get the selection list to appear.  I also just updated to iOS 16.2, so it may be that the Home App no longer renders this "off-use" combination.

If you can post a copy of what the HomeBridge database output look like, it should be replicable in HomeSpan.

**Reply by ghost** [reply]
Could` you try this? Works fine for me with 1.7.0 and iOS 16.2:

```
new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 

  new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("Light with Effects");
    new Service::LightBulb();
      new Characteristic::On();
      new Characteristic::Hue(0);
      new Characteristic::Saturation(0);
      new Characteristic::Brightness(100);

  SpanService *effect1 = new Service::InputSource();
    new Characteristic::ConfiguredName("Effect 1");
    new Characteristic::Identifier(1);

  SpanService *effect2 = new Service::InputSource();
    new Characteristic::ConfiguredName("Effect 2");
    new Characteristic::Identifier(2);

  (new Service::Television())
    ->addLink(effect1)
    ->addLink(effect2);
      new Characteristic::Active(1);
      new Characteristic::ActiveIdentifier(1); 
```

Btw, really love your library. Always wanted to get into this and it was never this easy!
One thing I noticed though:
Whenever I remove a bridge from HomeKit, I have to delete Device ID and Pairing Data (H command) before HomeKit can find the device again.

**Comment by HomeSpan** [comment]
Thanks very much for the code.  Based on the screen shot above, I had originally thought it might be possible to link InputSources into the LightBulb Service itself, but that did not work.   I had not thought of simply including the Television Service alongside the LightBulb Service within the same Accessory (I should have realized in the screen shot that there is an on/off button that looks like the one rendered for a Television Service, not the LightBulb Service).  In any case, really glad this works.

I have a Holiday Lights sketch that creates a variety of effects on a NeoPixel strip.  To select the effect I created a Custom Characteristic that can't be accessed by the Home App, but is nicely rendered in the Eve for HomeKit App (Eve seems to be able to "generically" render any Custom Characteristic).  Based on the above, I will try a second implementation of my Holiday Lights sketch that tacks on a Television Service so you can select the effects directly without needing to open the Eve for HomeKit App.

Re your question, I generally always type the H command before re-pairing if my configuration has changed.  This prevents the Home App from using any data it may have retained about the prior device with the same Device ID, which often leads it to get confused.  However, sometimes (though not always) if you leave it for a while, the Home App refreshes and updates to reflect the new device.

I've thought about resetting the Device ID every time the device is unpaired so that you don't need to type 'H', but this would be inconsistent with HAP specs since pairing/unpairing is not supposed to change the Device ID.   In practice, this seems to be just a convention, rather than a requirement.  I probably could add a function that would allow the user to automatically have HomeSpan run an 'H' command whenever unpaired.
